### PR TITLE
feat(appmesh): change HealthChecks to use protocol-specific union-like classes

### PR DESCRIPTION
**Pull request by dfezzie**
BREAKING CHANGE: HealthChecks require use of static factory methods

fixes https://github.com/aws/aws-cdk/issues/11640

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-appmesh/lib/private/utils.ts [hunk 1]
```diff
@@ -64,4 +22,14 @@ export interface ConnectionPoolConfig {
    * @default - none
    */
   readonly maxRequests?: number;
+}
+
+/**
+ * Enum of supported AppMesh protocols
+ */
+export enum Protocol {
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/private/utils.ts:30]: So, I moved this to the private section as it was not being used by any exported classes and this was giving a build error.

There is now an issue with the [ECS Service Extensions](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk-containers/ecs-service-extensions/lib/extensions/appmesh.ts#L33) project that was relying on this enum.

Should we move this back to the shared interfaces file and suppress the build warning, or add an enum specific to the ECS Service Extension package?

Comment (alexbrjo) [reply packages/@aws-cdk/aws-appmesh/lib/private/utils.ts:30]: if we added an enum specific to the ECS Service Extension package wouldn't we have to suppress the same warning?

File: packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts [hunk 2]
```diff
@@ -58,70 +59,160 @@ export interface TcpTimeout {
...
+    if (healthyThreshold < 2 || healthyThreshold > 10) {
+      throw new Error('healthyThreshold must be between 2 and 10');
+    }
+
+    if (unhealthyThreshold < 2 || unhealthyThreshold > 10) {
+      throw new Error('unhealthyThreshold must be between 2 and 10');
+    }
+
+    if (interval.toMilliseconds() < 5000 || interval.toMilliseconds() > 300_000) {
+      throw new Error('interval must be more than 5 seconds and less than 300 seconds');
```
Comment (alexbrjo) [inline packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:182]: nit: "at least 5 seconds and at most 300 seconds" :P

**Review by alexbrjo — APPROVED** [review]
Looks good, I like the option to set a default and great to see the code redundancy get reduced. Here's a consideration, it's possible to enforce different fields per protocol type without using factories. see the implementation of connection pooling: 

* [connection pool shapes](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-appmesh/lib/shared-interfaces.ts#L303-L356) that enforce the right fields at compile time 
* [the helper object](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-appmesh/lib/private/utils.ts#L43-L67) that allows any of these shapes get accepted by the rendering function
* the [L1 rendering function](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-appmesh/lib/virtual-node-listener.ts#L263-L271)

The con would be that using this approach would prevent us from setting defaults.

File: packages/@aws-cdk/aws-appmesh/lib/private/utils.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/private/utils.ts:30]: Looks like that module is stable. Given that, let's leave this enum as-is, but let's deprecate it. (`@deprecate` JSDoc tag)

File: packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts [hunk 3]
```diff
@@ -58,70 +59,160 @@ export interface TcpTimeout {
...
- * If property not set, then no healthchecks will be defined.
- */
-export interface HealthCheck {
+export interface HealthCheckCommonOptions {
   /**
-   * Number of successful attempts before considering the node UP
+   * The number of consecutive successful health checks that must occur before declaring listener healthy.
    *
-   * @default 2
+   * @default 5
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:68]: Why this change in the default, actually? Is this related to something in this PR, or just the docs were wrong previously?

File: packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts [hunk 4]
```diff
@@ -58,70 +59,160 @@ export interface TcpTimeout {
...
    */
   readonly healthyThreshold?: number;
 
   /**
-   * Interval in milliseconds to re-check
+   * The time period in milliseconds between each health check execution.
    *
-   * @default 5 seconds
+   * @default Duration.seconds(30)
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:75]: Same question here (why this change?).

File: packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts [hunk 5]
```diff
@@ -58,70 +59,160 @@ export interface TcpTimeout {
...
+   * The number of consecutive successful health checks that must occur before declaring listener healthy.
    *
-   * @default 2
+   * @default 5
    */
   readonly healthyThreshold?: number;
 
   /**
-   * Interval in milliseconds to re-check
+   * The time period in milliseconds between each health check execution.
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:73]: Let's get rid of the "milliseconds" here while we're changing these (the type here is `Duration`, no need to burden the customer with the details of what unit is required).

File: packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts [hunk 6]
```diff
@@ -58,70 +59,160 @@ export interface TcpTimeout {
...
    *
-   * @default 5 seconds
+   * @default Duration.seconds(30)
    */
   readonly interval?: cdk.Duration;
 
   /**
-   * The path where the application expects any health-checks, this can also be the application path.
+   * The amount of time to wait when receiving a response from the health check, in milliseconds.
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:80]: Same (remove milliseconds).

File: packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts [hunk 7]
```diff
@@ -58,70 +59,160 @@ export interface TcpTimeout {
...
+   * Construct a GRPC health check
+   */
+  public static grpc(options: HealthCheckCommonOptions= {}): HealthCheck {
+    return new HealthCheckImpl(Protocol.GRPC, options.healthyThreshold, options.unhealthyThreshold, options.interval, options.timeout);
+  }
+
+  /**
+   * Construct a TCP health check
+   */
+  public static tcp(options: HealthCheckCommonOptions = {}): HealthCheck {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:153]: Let's future proof these 2 a little bit:

```ts
// deliberately empty
export interface GrpcHealthCheckOptions extends HealthCheckCommonOptions {
}

export interface TcpHealthCheckOptions extends HealthCheckCommonOptions {
}

// ...

  public static grpc(options: GrpcHealthCheckOptions= {}): HealthCheck {
    // ...

  public static tcp(options: TcpHealthCheckOptions = {}): HealthCheck {
    // ...
```

File: packages/@aws-cdk/aws-appmesh/lib/virtual-gateway-listener.ts [hunk 8]
```diff
@@ -143,7 +141,7 @@ class VirtualGatewayListenerImpl extends VirtualGatewayListener {
           port: this.port,
           protocol: this.protocol,
         },
-        healthCheck: this.healthCheck ? renderHealthCheck(this.healthCheck, this.protocol, this.port): undefined,
+        healthCheck: this.healthCheck ? this.healthCheck.bind(scope).virtualGatewayHealthCheck: undefined,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-gateway-listener.ts:144]: You can save a few characters here by using `?.`:

```suggestion
        healthCheck: this.healthCheck?.bind(scope).virtualGatewayHealthCheck,
```

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts [hunk 9]
```diff
@@ -185,7 +184,7 @@ class VirtualNodeListenerImpl extends VirtualNodeListener {
           port: this.port,
           protocol: this.protocol,
         },
-        healthCheck: this.healthCheck ? this.renderHealthCheck(this.healthCheck) : undefined,
+        healthCheck: this.healthCheck ? this.healthCheck.bind(scope).virtualNodeHealthCheck : undefined,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts:187]: Same:

```suggestion
        healthCheck: this.healthCheck?.bind(scope).virtualNodeHealthCheck,
```

File: packages/@aws-cdk/aws-appmesh/test/integ.mesh.expected.json [hunk 10]
```diff
@@ -1270,12 +1267,11 @@
...
               "HealthCheck": {
-                "HealthyThreshold": 2,
+                "HealthyThreshold": 5,
                 "IntervalMillis": 10000,
                 "Path": "/",
-                "Port": 443,
                 "Protocol": "http",
-                "TimeoutMillis": 2000,
+                "TimeoutMillis": 5000,
                 "UnhealthyThreshold": 2
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/integ.mesh.expected.json:1275]: I'm worried about these changes...? What happened here?

File: packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts [hunk 11]
```diff
@@ -67,15 +67,13 @@ router.addRoute('route-1', {
 const node2 = mesh.addVirtualNode('node2', {
   serviceDiscovery: appmesh.ServiceDiscovery.dns(`node2.${namespace.namespaceName}`),
   listeners: [appmesh.VirtualNodeListener.http({
-    healthCheck: {
+    healthCheck: appmesh.HealthCheck.http({
       healthyThreshold: 3,
       interval: cdk.Duration.seconds(5),
       path: '/check-path2',
-      port: 8080,
-      protocol: appmesh.Protocol.HTTP,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts:75]: Why these changes? We really should strive to keep changes to tests to a minimum when doing these kinds of refactorings.

File: packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts [hunk 12]
```diff
@@ -93,15 +91,13 @@ const node2 = mesh.addVirtualNode('node2', {
 const node3 = mesh.addVirtualNode('node3', {
   serviceDiscovery: appmesh.ServiceDiscovery.dns(`node3.${namespace.namespaceName}`),
   listeners: [appmesh.VirtualNodeListener.http({
-    healthCheck: {
+    healthCheck: appmesh.HealthCheck.http({
       healthyThreshold: 3,
       interval: cdk.Duration.seconds(5),
       path: '/check-path3',
-      port: 8080,
-      protocol: appmesh.Protocol.HTTP,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts:101]: Same comment here (let's minimize changes to tests with this PR).

File: packages/@aws-cdk/aws-appmesh/test/test.mesh.ts [hunk 13]
```diff
@@ -236,7 +236,6 @@ export = {
                     HealthyThreshold: 3,
                     IntervalMillis: 5000,
                     Path: '/',
-                    Port: 8080,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.mesh.ts:239]: Again, this change is concerning to me. Why are we changing this?

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts [hunk 14]
```diff
@@ -59,17 +58,16 @@ export = {
...
               HealthCheck: {
-                HealthyThreshold: 2,
-                IntervalMillis: 10000,
-                Port: 443,
-                Protocol: appmesh.Protocol.HTTP,
-                TimeoutMillis: 2000,
+                HealthyThreshold: 5,
+                IntervalMillis: 10_000,
+                Protocol: Protocol.HTTP,
+                TimeoutMillis: 5000,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts:64]: Again, test changes where there shouldn't be any.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts [hunk 15]
```diff
@@ -82,17 +80,16 @@ export = {
...
               HealthCheck: {
-                HealthyThreshold: 2,
-                IntervalMillis: 10000,
-                Port: 443,
-                Protocol: appmesh.Protocol.HTTP2,
-                TimeoutMillis: 2000,
+                HealthyThreshold: 5,
+                IntervalMillis: 10_000,
+                Protocol: Protocol.HTTP2,
+                TimeoutMillis: 5000,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts:86]: Same here (test changes?).

**Review by skinny85 — DISMISSED** [review]
Thanks for the PR @dfezzie!

I have some comments. I'm mainly concerned with the changes to the defaults, which result in a lot of test changes. I really think we should strive to minimize changes to tests when mixing them with functional changes.

Is there any way to preserve the old values in the templates, while changing the API?

File: packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts (hunk 5 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:73]: Good catch. Will do

File: packages/@aws-cdk/aws-appmesh/test/test.mesh.ts (hunk 13 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/test/test.mesh.ts:239]: `port` was dropped from the interface as we only allow it to be the same as the listener, which is the default when port is not specified. Seems like a bad customer experience to have an option to configure it

File: packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts (hunk 4 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:86]: We can change these back if you prefer -- I just don't know how they were originally chosen.

Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:86]: @dfezzie I'm a little wary of changing so many things at the same time in a single PR. Is there any way we can split this PR in 2 then?

1. In the first PR, change the API to use the union-like pattern, while keeping the defaults the same.
2. In the second PR, change the defaults.

Or do you think that's too irritating/too much work?

Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:86]: I'm fine with that!

Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:86]: I'd like to keep dropping the port from the interface though. I'll revert the defaults though

**Comment by dfezzie** [comment]
Hey @skinny85 I've finished the update here. Let me know if there are any other changes I need to add!

File: packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts [hunk 16]
```diff
@@ -68,60 +70,170 @@ export enum Protocol {
...
+   * The destination path for the health check request.
    *
-   * @default - same as corresponding port mapping
+   * @default /
    */
-  readonly protocol?: Protocol;
+  readonly path?: string;
+}
+
+/**
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:117]: Can we move all of the newly created classes and interfaces into their own, separate file?

File: packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts [hunk 17]
```diff
@@ -68,60 +70,170 @@ export enum Protocol {
...
+      },
+      virtualGatewayHealthCheck: {
+        protocol: this.protocol,
+        healthyThreshold: this.healthyThreshold,
+        unhealthyThreshold: this.unhealthyThreshold,
+        intervalMillis: this.interval.toMilliseconds(),
+        timeoutMillis: this.timeout.toMilliseconds(),
+        path: this.path,
+      },
+    };
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:234]: So, this results in dropping the `port` property from the `healthCheck` property of the currently generated CloudFormation templates. Meaning, customers who update their CDK version will see a difference from their previous template when they run `cdk diff`, even though there's no functional change here.

Can I suggest a simple trick that will make this difference go away, without exposing `port` as an input property?

Let's tweak the `HealthCheck` API slightly:

```ts
export interface HealthCheckBindOptions {
  readonly defaultPort?: number;
}

export abstract class HealthCheck
  // ...

  public abstract bind(scope: Construct, options: HealthCheckBindOptions): HealthCheckConfig;
}
```

And then, in `HealthCheckImpl`:

```ts
class HealthCheckImpl extends HealthCheck {
  // ...

  public bind(_scope: Construct, options: HealthCheckBindOptions): HealthCheckConfig {
    return {
      virtualNodeHealthCheck: {
        // remaining properties same as in the PR...
        port: options.port,
      },
      virtualGatewayHealthCheck: {
        // remaining properties same as in the PR...
        port: options.port,
      },
    };
```

And then `bind()` will be called as:

```ts
        healthCheck: this.healthCheck?.bind(scope, { defaultPort: this.port }).virtualGatewayHealthCheck,
```

What do you think @dfezzie?

**Review by skinny85 — COMMENTED** [review]
Looks great @dfezzie!

Two small suggestions, let me know what you think!

File: packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts (hunk 17 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:234]: I like that. Makes sense! I'll make the update

**Review by skinny85 — APPROVED** [review]
Looks awesome @dfezzie! 👍

**Review by skinny85 — APPROVED** [review]
